containerd-shim-v2/service.go

```diff
@@ -760,20 +823,22 @@ func (s *service) processExits() {
 }
 
 func (s *service) checkProcesses(e exit) {
-	s.Lock()
-	defer s.Unlock()
+	s.mu.Lock()
+	defer s.mu.Unlock()
 
 	id := e.execid
 	if id == "" {
 		id = e.id
 	}
-	s.events <- &eventstypes.TaskExit{
+
+	s.sendL(&eventstypes.TaskExit{
```

Comment:
Another confused me is there is two function for send event, `send`(// for unit test, it will not initialize s.events) and `sendL`,  `sendL` is only used here, so that means `Start` and `checkProcesses` need lock, but other function do not need, a little confused.

Reply:
Yes, here the lock is mainly used for preventing the case of start events are sent before any exit events in the error case.
For example, shimv2 start a container, but for some reasons the start failed and the container process exit immediately, without this lock, we cannot guarantee which event send first, the create event or the exit event? So this lock is used to make sure the create event is send first.